### PR TITLE
Add compat random wrappers

### PR DESCRIPTION
**Pull request by lsm**
## Summary
- Add compat random helper boundaries for secure vs ordinary random byte generation, plus a deterministic test source.
- Route PKCE verifier generation, WebSocket masks/nonces, and provider protocol IDs through the compat random seam.
- Add deterministic PKCE seam tests and include the utils PKCE module in the build test graph.

Part of stack: Zig 0.15.2 → 0.16.0 Migration. PR 4 of 24.

## Test plan
- [x] ./scripts/check-zig-patterns.sh
- [x] cd zig && /tmp/zig-x86_64-macos-0.15.2/zig build test-unit-utils
- [x] cd zig && /tmp/zig-x86_64-macos-0.15.2/zig build test-unit-protocol
- [x] cd zig && /tmp/zig-x86_64-macos-0.15.2/zig build test-unit-transport
- [x] cd zig && /tmp/zig-x86_64-macos-0.15.2/zig build test

Note: the default `zig` on this runner is 0.16.0 and fails current 0.15.2-targeted code before this PR's changes; tests were run with the local Zig 0.15.2 binary.

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: REQUEST_CHANGES

### Summary

The PR implements the random wrapper compat layer well overall: clean API surface with `fillSecureBytes`/`fillRandomBytes` (fill-style) and `secureBytes`/`randomBytes` (allocating), a `DeterministicSource` for test seams, proper PKCE routing through `fillSecureBytes`, WebSocket nonce/mask routing, ULID random part routing, and correct build graph wiring for the previously-missing `oauth_utils_pkce_test`. Tests are solid (length, non-empty, uniqueness, determinism, empty buffers).

However, there is one **correctness regression** that needs fixing before this ships.

---

### P1 — Modulo bias regression in `generateSessionId()`

**File:** `zig/src/protocol/provider/types.zig`, lines 33–36

The original code used `std.crypto.random.intRangeLessThan(usize, 0, SESSION_ID_ALPHABET.len)`, which internally uses **rejection sampling** to guarantee a perfectly uniform distribution over the 62-character alphabet.

The replacement:
```zig
var random_index: [2]u8 = undefined;
compat.random.fillRandomBytes(&random_index);
const raw_index = std.mem.readInt(u16, &random_index, .little);
const idx = raw_index % SESSION_ID_ALPHABET.len;
```

introduces **modulo bias**: 65536 ÷ 62 = 1057 remainder 2, so `'0'` and `'1'` each appear ~0.095% more often than the other 60 characters. While negligible for session IDs, this is a **regression** — the original was bias-free.

**Fix options:**
1. Add a `fillRandomIntRangeLessThan(comptime T: type, min: T, max: T) T` (or similar) helper to `compat/random.zig` that wraps rejection sampling and delegates to `fillRandomBytes` internally. This would also help when migrating `storage.zig` (which uses `std.crypto.random.int(u64)`).
2. Inline a rejection-sampling loop in `generateSessionId`:
   ```zig
   var random_index: [2]u8 = undefined;
   var idx: usize = 0;
   while (true) {
       compat.random.fillRandomBytes(&random_index);
       const raw = std.mem.readInt(u16, &random_index, .little);
       // 65536 - 65536%62 = 65534; reject values >= 65534 to eliminate bias
       if (raw < 65536 - (65536 % SESSION_ID_ALPHABET.len)) {
           idx = raw % SESSION_ID_ALPHABET.len;
           break;
       }
   };
   ```

---

### P3 — Two `fillTestPkceBytes` implementations are identical

**Files:** `zig/src/oauth/pkce.zig:52–55` and `zig/src/utils/oauth/pkce.zig:50–53`

Both PKCE modules contain an identical `fillTestPkceBytes` helper. The task description explicitly notes these two modules "should either share the secure helper or be deduplicated in a follow-up." No action required now, but noting for the follow-up.

---

### Items verified ✓

- ✅ Wrappers compile on Zig 0.15.2, all tests pass (`test-unit-utils`, `test-unit-protocol`, `test-unit-transport`, full `test`)
- ✅ `check-zig-patterns.sh` NOT modified to fail on direct `std.crypto.random` (correct for Phase 0)
- ✅ Security-sensitive paths (PKCE, WebSocket mask/nonce) use `fillSecureBytes`
- ✅ Non-security paths (ULID randomness, session IDs) use `fillRandomBytes`
- ✅ `DeterministicSource` is well-designed with `bytes()` and `allocBytes()`
- ✅ `oauth_utils_pkce_test` properly wired into build graph (was previously missing)
- ✅ No `std.Io` in public API signatures
- ✅ 0.16 migration comments present
- ✅ Five remaining `std.crypto.random` direct uses (openai_codex, google_gemini_cli, google_antigravity, tool_utils, storage) are correctly left for follow-up per scope

File: zig/src/protocol/provider/types.zig [hunk 1]
```diff
@@ -30,7 +30,10 @@ pub const SessionId = [SESSION_ID_LENGTH]u8;
 pub fn generateSessionId() SessionId {
     var session_id: SessionId = undefined;
     for (&session_id) |*byte| {
-        const idx = std.crypto.random.intRangeLessThan(usize, 0, SESSION_ID_ALPHABET.len);
```
Comment (lsm) [inline zig/src/protocol/provider/types.zig:33]: **P1 — Modulo bias regression.** The original `std.crypto.random.intRangeLessThan` used rejection sampling for uniform distribution. The replacement `u16 % 62` introduces bias: 65536 ÷ 62 = 1057 r 2, so `'0'` and `'1'` appear ~0.095% more often. Suggest adding a rejection-sampling wrapper to `compat/random.zig` (e.g., `fillRandomIntRangeLessThan`) or inlining the rejection loop here.

Comment (lsm) [reply zig/src/protocol/provider/types.zig:33]: Fixed in 9756d81 by adding compat random range helpers that delegate to std.crypto.random.intRangeLessThan (preserving rejection-sampling semantics on 0.15.2) and switching generateSessionId() to use the secure unbiased range helper instead of modulo.

File: zig/src/protocol/provider/types.zig [hunk 2]
```diff
@@ -30,7 +30,10 @@ pub const SessionId = [SESSION_ID_LENGTH]u8;
...
+        var random_index: [2]u8 = undefined;
+        compat.random.fillRandomBytes(&random_index);
+        const raw_index = std.mem.readInt(u16, &random_index, .little);
+        const idx = raw_index % SESSION_ID_ALPHABET.len;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/protocol/provider/types.zig]: Fixed in 9756d81. generateSessionId() no longer maps random bytes with %, and now uses compat.random.secureIntRangeLessThan(), which preserves std.crypto.random.intRangeLessThan rejection-sampling behavior.

File: zig/src/protocol/provider/types.zig [hunk 3]
```diff
@@ -30,7 +30,10 @@ pub const SessionId = [SESSION_ID_LENGTH]u8;
 pub fn generateSessionId() SessionId {
     var session_id: SessionId = undefined;
     for (&session_id) |*byte| {
-        const idx = std.crypto.random.intRangeLessThan(usize, 0, SESSION_ID_ALPHABET.len);
+        var random_index: [2]u8 = undefined;
+        compat.random.fillRandomBytes(&random_index);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/protocol/provider/types.zig]: Fixed in 9756d81. generateSessionId() now uses the secure random range seam so it keeps the previous secure entropy posture across the 0.16 remap.

File: zig/src/protocol/provider/types.zig [hunk 4]
```diff
@@ -98,7 +101,7 @@ pub fn generateUlid() Ulid {
     ulid[3] = @intCast((now_ms >> 16) & 0xff);
     ulid[4] = @intCast((now_ms >> 8) & 0xff);
     ulid[5] = @intCast(now_ms & 0xff);
-    std.crypto.random.bytes(ulid[6..16]);
+    compat.random.fillRandomBytes(ulid[6..16]);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/protocol/provider/types.zig]: Fixed in 9756d81. ULID random bytes now use compat.random.fillSecureBytes() so protocol stream/message IDs stay on the secure entropy seam.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: APPROVE

### Summary

The P1 modulo bias regression has been fixed. `generateSessionId()` now uses `secureIntRangeLessThan(usize, SESSION_ID_ALPHABET.len)` which preserves the rejection-sampling semantics of the original `std.crypto.random.intRangeLessThan`. The `generateUlid()` path correctly uses `fillSecureBytes`.

### What was addressed

- **P1 fixed:** New `secureIntRangeLessThan` and `randomIntRangeLessThan` helpers added to `compat/random.zig` — both delegate to `std.crypto.random.intRangeLessThan` with rejection sampling, no modulo bias.
- `generateSessionId()` switched to `secureIntRangeLessThan` — bias-free.
- ULID random bytes switched to `fillSecureBytes` — consistent security semantics for protocol stream/message IDs.
- New tests for range helpers: upper bound respect (128 iterations), single-value range edge case.

### Items verified ✓

- ✅ All tests pass: `test-unit-utils`, `test-unit-protocol`, `test-unit-transport`, full `test`
- ✅ No modulo bias — `secureIntRangeLessThan` wraps `intRangeLessThan` (rejection sampling)
- ✅ Only `compat/random.zig` touches `std.crypto.random` directly — 5 remaining direct uses are correctly left for follow-up per scope
- ✅ All review threads resolved
- ✅ `check-zig-patterns.sh` not modified to fail on direct `std.crypto.random`
- ✅ No `std.Io` in public API signatures
- ✅ 0.16 migration comments present on all new helpers

File: zig/src/oauth/pkce.zig [hunk 5]
```diff
@@ -1,4 +1,5 @@
 const std = @import("std");
+const compat = @import("compat");
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/oauth/pkce.zig:2]: Fixed in 0aa8a2c by adding the compat module import to the oauth_test root module, so src/oauth/pkce.zig can resolve @import("compat") when compiled through src/oauth/mod.zig.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: APPROVE

### Summary

New commit `0aa8a2c` adds the missing `compat` import to the `oauth_test` root module in `build.zig`, fixing the import resolution path for `src/oauth/pkce.zig` when compiled through `src/oauth/mod.zig`. One-line fix, correct and minimal.

All prior-round findings (P1 modulo bias) remain addressed. All tests pass (`test-unit-utils`, full `test`). All review threads resolved. PR is clean and mergeable.
